File: src/lib/teams.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Wire types mirroring `orchestrator::teams::{Team, Worker}`. A "team" is a
 * named coordination unit persisted at `~/.cortex/teams/<id>.json` and
 * surfaced in the orchestrator dashboard.
 *
 * Status pills render off `WorkerStatus`; any string outside the union is
 * rejected at write-time on the Rust side so the renderer can be exhaustive.
 */

export type WorkerStatus = "idle" | "working" | "blocked" | "done" | "error";

export const WORKER_STATUSES: WorkerStatus[] = [
  "idle",
  "working",
  "blocked",
  "done",
  "error",
];

/** Team-level run lifecycle (see `Team.run_status`). */
export type TeamRunStatus = "planning" | "running" | "done" | "error";

export interface Worker {
  role: string;
  agent_id: string;
  current_task?: string | null;
  status: WorkerStatus | string;
  started_unix_ms: number;
  last_event_unix_ms: number;
  message_count: number;
  /** Chat session holding this worker's latest run transcript. */
  session_id?: string | null;
  /**
   * Subtask classification from the manager's plan (orchestration slice 2):
   * `"chat" | "code"` and `"easy" | "medium" | "hard"`. Absent until a run tags
   * the worker; cost-aware dispatch (slice 3) reads them.
   */
  task_kind?: string | null;
  task_difficulty?: string | null;
  /**
   * Cost-aware dispatch record (slice 3). `effective_model` is the concrete
   * model this worker ran on (a role pin or the cost router's pick);
   * `projected_usd` is the estimated spend for its latest run (0 for free local
   * models). Absent until a run dispatches the worker.
   */
  effective_model?: string | null;
  projected_usd?: number | null;
  /**
   * The `lane_runs` id a `code`-tagged worker's subtask was dispatched to
   * (orchestration slice 4): with a repo bound, code work edits it in a gateway
   * worktree lane instead of producing chat. Absent for chat workers. Deep-links
   * the run into the Lanes tab.
   */
  lane_run_id?: string | null;
}

export interface Team {
  id: string;
  name: string;
  manager_role: string;
  workers: Worker[];
  created_unix_ms: number;
  /** The goal handed to the manager on the most recent run. */
  goal?: string | null;
  /** Lifecycle of the most recent run; absent = never run. */
  run_status?: TeamRunStatus | string | null;
  last_run_unix_ms?: number | null;
  /** Chat session holding the manager's planning transcript. */
  plan_session_id?: string | null;
  /**
   * Total estimated USD across the most recent run's workers (slice 3). Absent
   * until a run completes; `0` is legitimate (an all-local run is free).
   */
  spent_usd?: number | null;
  /**
   * Chat session holding the manager's synthesis + verification pass over all
   * the workers' outputs (slice 5). Absent for single-worker / trivially-thin
   * runs (the synthesis gate) and until a multi-worker run completes.
   */
  synthesis_session_id?: string | null;
  /**
   * Optional soft spend ceiling in USD (slice 6). When set, the dashboard
   * compares the run's projected `spent_usd` against it and paints an
   * over-budget warning if exceeded — it never blocks or kills a run. Absent =
   * no budget. A team-level setting (survives runs, unlike `spent_usd`).
   */
  budget_usd?: number | null;
}

/** True while the team's most recent run is still in flight. */
export function isTeamRunning(team: Team): boolean {
  return team.run_status === "planning" || team.run_status === "running";
}

/** List every team under `~/.cortex/teams/*.json`, newest first. */
export async function listTeams(): Promise<Team[]> {
  return invoke<Team[]>("list_teams");
}

/** Load a single team by id. Throws on missing / malformed. */
export async function getTeam(id: string): Promise<Team> {
  return invoke<Team>("get_team", { id });
}

/**
 * Create + persist a new team. Workers all start in `idle`. `budgetUsd` is an
 * optional soft spend ceiling (slice 6); `null` = no budget.
 */
export async function createTeam(
  name: string,
  managerRole: string,
  workerRoles: string[],
  budgetUsd: number | null = null,
): Promise<Team> {
  return invoke<Team>("create_team", {
    name,
    managerRole,
    workerRoles,
    budgetUsd,
  });
}

/**
 * Mutate one worker. `status` must be a `WorkerStatus`. `currentTask = null`
 * leaves the existing task untouched; pass a string (including `""`) to set.
 */
export async function updateTeamWorker(
  teamId: string,
  workerId: string,
  status: WorkerStatus,
  currentTask: string | null = null,
): Promise<Team> {
  return invoke<Team>("update_team_worker", {
    teamId,
    workerId,
    status,
    currentTask,
  });
}

/**
 * Kick off a real team run: the manager plans one task per worker, then the
 * workers execute concurrently through the adapter registry. Resolves as soon
 * as the run is accepted (team comes back stamped `planning`); progress lands
 * in the team file — follow it via polling or the `teams:updated` event.
 */
export async function runTeam(
  teamId: string,
  goal: string,
  model: string | null = null,
  repo: string | null = null,
  budgetUsd: number | null = null,
): Promise<Team> {
  return invoke<Team>("run_team", { teamId, goal, model, repo, budgetUsd });
}

/** Remove a team file. Missing files are a no-op. */
export async function deleteTeam(id: string): Promise<void> {
  return invoke("delete_team", { id });
}

/** Human-friendly "5m ago" rendering for `*_unix_ms` columns. */
export { timeAgo } from "@/lib/time";

/** Short truncate for the per-card task line — keeps cards uniform height. */
export function truncateTask(task: string | null | undefined, max = 90): string {
  if (!task) return "—";
  const t = task.trim();
  if (t.length <= max) return t;
  return `${t.slice(0, max - 1)}…`;
}

/**
 * Format an estimated USD spend for the cost-aware dispatch surface (slice 3).
 * Free (all-local) runs read "Free" rather than "$0.00"; sub-cent amounts keep
 * enough precision to stay non-zero so a real-but-tiny estimate isn't hidden.
 */
export function formatUsd(usd: number | null | undefined): string {
  if (usd == null) return "—";
  if (usd <= 0) return "Free";
  if (usd < 0.01) return `<$0.01`;
  return `$${usd.toFixed(2)}`;
}

export interface BudgetStatus {
  /** The team's spend ceiling. */
  budget: number;
  /** The run's projected spend (0 when no run has completed yet). */
  projected: number;
  /** projected / budget clamped to [0, 1] for the gauge fill width. */
  ratio: number;
  /** True when the projection has crossed the budget (the soft-warn trigger). */
  over: boolean;
}

/**
 * Compare a team's projected spend against its soft budget (slice 6). Returns
 * `null` when no budget is set (nothing to gauge). A zero budget with any spend
 * reads as over; a zero budget with zero spend is exactly at-budget (not over),
 * so an all-free run under a $0 ceiling stays calm.
 */
export function budgetStatus(
  budget: number | null | undefined,
  spent: number | null | undefined,
): BudgetStatus | null {
  if (budget == null) return null;
  const projected = spent ?? 0;
  const ratio = budget > 0 ? Math.min(1, Math.max(0, projected / budget)) : projected > 0 ? 1 : 0;
  return { budget, projected, ratio, over: projected > budget };
}

File: src-tauri/src/commands/teams.rs
//! Tauri commands backing `~/.cortex/teams/*.json` — multi-agent team state
//! consumed by the orchestrator dashboard. All heavy lifting (validation,
//! disk IO, id minting) lives in `orchestrator::teams`; this file just adapts
//! the wire shape to async Tauri handlers and converts `anyhow::Error` to the
//! frontend-friendly `Result<_, String>` form.

use crate::app_state::AppState;
use crate::lanes::TERMINAL_STATUSES;
use crate::observability::tracing_store::TracingStore;
use crate::orchestrator::team_run::{self, LaneDispatch, LaneDispatcher};
use crate::orchestrator::teams::{self, Team};
use std::sync::Arc;
use std::time::{Duration, Instant};
use tauri::{Emitter, Manager, State};

/// Cap on how long a single team code-lane may run before the worker is freed
/// (the lane keeps running on the gateway; the row stays in the Lanes tab). Matches
/// the chat worker's wall-clock budget in `team_run`.
const LANE_FOLLOW_TIMEOUT: Duration = Duration::from_secs(420);
/// Poll cadence while following a lane to a terminal state.
const LANE_POLL_INTERVAL: Duration = Duration::from_millis(400);

/// Bridges the Tauri-free team engine to the Lanes machinery (slice 4): a
/// `Code`-tagged worker's subtask is dispatched as a gateway worktree lane via
/// [`crate::commands::multi_provider::dispatch_team_lane`], then followed to a
/// terminal state by observing the `lane_runs` row the watcher updates.
struct TeamLaneDispatcher {
    app: tauri::AppHandle,
    owner: String,
    repo: String,
    provider: String,
}

#[async_trait::async_trait]
impl LaneDispatcher for TeamLaneDispatcher {
    async fn dispatch(&self, goal: &str, task: &str) -> Result<LaneDispatch, String> {
        let instructions = format!("Team goal: {goal}");
        let record = crate::commands::multi_provider::dispatch_team_lane(
            &self.app,
            &self.owner,
            &self.repo,
            &self.provider,
            task,
            Some(&instructions),
        )
        .await?;
        let lane_run_id = record.run_id.clone();
        let branch = record
            .branch
            .as_deref()
            .map(|b| format!("Lane branch `{b}`. "))
            .unwrap_or_default();

        // A lane that failed to even start is already terminal.
        if record.status != "running" {
            return Ok(LaneDispatch {
                lane_run_id,
                status: record.status.clone(),
                outcome: format!(
                    "{branch}{}",
                    record.detail.as_deref().unwrap_or("the lane did not start")
                ),
                provider: self.provider.clone(),
            });
        }

        // Follow the lane to a terminal state by observing the row the detached
        // watcher updates (single source of truth — no second SSE consumer).
        let store = crate::commands::multi_provider::lane_store(&self.app);
        let deadline = Instant::now() + LANE_FOLLOW_TIMEOUT;
        loop {
            if Instant::now() >= deadline {
                return Ok(LaneDispatch {
                    lane_run_id,
                    status: "error".into(),
                    outcome: format!(
                        "{branch}still running after {}s — follow it in the Lanes tab.",
                        LANE_FOLLOW_TIMEOUT.as_secs()
                    ),
                    provider: self.provider.clone(),
                });
            }
            tokio::time::sleep(LANE_POLL_INTERVAL).await;
            let row = store
                .get(&lane_run_id)
                .map_err(|e| format!("lane lookup failed: {e}"))?;
            if let Some(r) = row {
                if TERMINAL_STATUSES.contains(&r.status.as_str()) {
                    return Ok(LaneDispatch {
                        lane_run_id,
                        status: r.status.clone(),
                        outcome: format!("{branch}{}", r.detail.as_deref().unwrap_or("completed")),
                        provider: self.provider.clone(),
                    });
                }
            }
        }
    }
}

/// Build a lane dispatcher for a run when a repo is bound (slice 4). `None` —
/// no repo, or an unparseable `owner/repo` — keeps code workers on the chat
/// path. Under `CORTEX_E2E` the provider is forced to the deterministic
/// `e2e-fake` lane producer so the probe never dials the gateway.
fn build_lane_dispatcher(
    app: &tauri::AppHandle,
    repo: Option<&str>,
    model: Option<&str>,
) -> Option<Arc<dyn LaneDispatcher>> {
    let (owner, repo) = parse_owner_repo(repo?)?;
    let provider = if crate::commands::e2e::e2e_enabled() {
        "e2e-fake".to_string()
    } else {
        model
            .map(str::trim)
            .filter(|s| !s.is_empty())
            .unwrap_or("claude")
            .to_string()
    };
    Some(Arc::new(TeamLaneDispatcher {
        app: app.clone(),
        owner,
        repo,
        provider,
    }))
}

/// Parse a Gitea `owner/repo` (optional `.git` suffix) into its two parts.
/// Returns `None` for anything that isn't exactly one `/`-separated pair.
fn parse_owner_repo(raw: &str) -> Option<(String, String)> {
    let cleaned = raw.trim().trim_end_matches(".git");
    let (owner, repo) = cleaned.split_once('/')?;
    let (owner, repo) = (owner.trim(), repo.trim());
    if owner.is_empty() || repo.is_empty() || repo.contains('/') {
        return None;
    }
    Some((owner.to_string(), repo.to_string()))
}

/// List every team under `~/.cortex/teams/*.json`, newest first. Missing
/// directory yields `[]` — never an error.
#[tauri::command]
pub async fn list_teams() -> Result<Vec<Team>, String> {
    Ok(teams::list_teams())
}

/// Load a single team by id. Errors if the file is missing or malformed.
#[tauri::command]
pub async fn get_team(id: String) -> Result<Team, String> {
    if id.trim().is_empty() {
        return Err("id is required".into());
    }
    teams::get_team(&id).ok_or_else(|| format!("team '{id}' not found"))
}

/// Create + persist a new team. Workers all start in `idle` with no task.
/// `budget_usd` is an optional soft spend ceiling (slice 6) — when present it's
/// stamped onto the new team so the dashboard can flag over-budget projections.
#[tauri::command]
pub async fn create_team(
    name: String,
    manager_role: String,
    worker_roles: Vec<String>,
    budget_usd: Option<f64>,
) -> Result<Team, String> {
    let team =
        teams::create_team(&name, &manager_role, &worker_roles).map_err(|e| e.to_string())?;
    match budget_usd {
        Some(_) => teams::set_budget_usd(&team.id, budget_usd).map_err(|e| e.to_string()),
        None => Ok(team),
    }
}

/// Mutate a single worker's status / current_task on `team_id`. Returns the
/// updated team so the UI can re-render without a follow-up `get_team` call.
#[tauri::command]
pub async fn update_team_worker(
    team_id: String,
    worker_id: String,
    status: String,
    current_task: Option<String>,
) -> Result<Team, String> {
    teams::update_worker(&team_id, &worker_id, &status, current_task)
        .map_err(|e| e.to_string())
}

/// Remove a team file. Missing files are a no-op.
#[tauri::command]
pub async fn delete_team(id: String) -> Result<(), String> {
    if id.trim().is_empty() {
        return Err("id is required".into());
    }
    if team_run::is_running(&id) {
        return Err("That team is mid-run — wait for it to finish before deleting.".into());
    }
    teams::delete_team(&id).map_err(|e| e.to_string())
}

/// Kick off a real team run: the manager plans one task per worker, then the
/// workers execute concurrently through the adapter registry (see
/// `orchestrator::team_run`). Returns the team already stamped `planning` —
/// the run itself continues detached (it survives tab switches; the dashboard
/// follows along via its poll + the `teams:updated` event). `model` is the
/// composer's current pick; `None` falls back to role models / default route.
#[tauri::command]
pub async fn run_team(
    team_id: String,
    goal: String,
    model: Option<String>,
    repo: Option<String>,
    budget_usd: Option<f64>,
    app: tauri::AppHandle,
    state: State<'_, AppState>,
) -> Result<Team, String> {
    if goal.trim().is_empty() {
        return Err("Give the team a goal first.".into());
    }
    if !team_run::try_begin(&team_id) {
        return Err("That team is already running — one goal at a time.".into());
    }
    // Slice 6: the Assign-goal modal can set/adjust this run's soft budget. Apply
    // it before stamping `planning` so the returned team carries the new ceiling;
    // a bad value releases the guard rather than wedging the team.
    if budget_usd.is_some() {
        if let Err(e) = teams::set_budget_usd(&team_id, budget_usd) {
            team_run::finish(&team_id);
            return Err(e.to_string());
        }
    }
    // Stamp `planning` BEFORE returning so the UI reflects the run instantly;
    // on failure release the guard or it would wedge the team until restart.
    let team = match teams::begin_run(&team_id, &goal) {
        Ok(t) => t,
        Err(e) => {
            team_run::finish(&team_id);
            return Err(e.to_string());
        }
    };
    let _ = app.emit("teams:updated", &team_id);

    let registry = state.registry.clone();
    let store = app.state::<TracingStore>().inner().clone();
    // Slice 4: when a repo is bound, code-tagged subtasks edit it in a gateway
    // worktree lane; otherwise code workers stay on the chat path.
    let lane_dispatcher = build_lane_dispatcher(&app, repo.as_deref(), model.as_deref());
    tauri::async_runtime::spawn(async move {
        let notify_app = app.clone();
        let notify_id = team_id.clone();
        let result = team_run::execute_team(
            registry,
            store,
            team_id.clone(),
            goal,
            model,
            lane_dispatcher,
            move || {
                let _ = notify_app.emit("teams:updated", &notify_id);
            },
        )
        .await;
        if let Err(e) = result {
            tracing::warn!("team run {team_id} failed: {e}");
        }
        team_run::finish(&team_id);
    });
    Ok(team)
}

#[cfg(test)]
mod tests {
    use super::parse_owner_repo;

    #[test]
    fn parses_owner_repo_pairs() {
        assert_eq!(
            parse_owner_repo("NetworkBound/cortex"),
            Some(("NetworkBound".into(), "cortex".into()))
        );
        // Trims and strips a .git suffix.
        assert_eq!(
            parse_owner_repo("  owner/repo.git "),
            Some(("owner".into(), "repo".into()))
        );
    }

    #[test]
    fn rejects_malformed_repo_specs() {
        for bad in ["", "noslash", "a/b/c", "/repo", "owner/", "   /   "] {
            assert_eq!(parse_owner_repo(bad), None, "{bad:?} should not parse");
        }
    }
}
